### ISSUE TITLE
Bug: 

### ISSUE THREAD

**Issue by lukigarazus** [facebook/react#23040]
React version: 17.0.2

## Steps To Reproduce

1. Open the below page
2. Open the console and see "2" logged out 2 times
3. Remove the line under the "remove this line" comment
4. Reload the app
5. Open the console and see "2" and "1" logged out

Link to code example: https://codesandbox.io/s/angry-curie-ln5u0?file=/src/App.tsx

## The current behavior
"setState" calls are not batched properly in async/await functions.

## The expected behavior
"setState" calls should be batched correctly in all contexts.

**Comment by anton-g** [comment]
Unless I'm mistaken this is how batching works in React 17 and earlier. In React 18 however, Automatic Batching have been added. Here's your sandbox but running on the latest RC for 18: https://codesandbox.io/s/optimistic-lake-x0zfp

You can read more about batching in React 18 here: https://github.com/reactwg/react-18/discussions/21

In <17 there's also `ReactDOM.unstable_batchedUpdates` that you can use to wrap your setStates and force React to batch them, but as the unstable prefix hints about it's not really recommended to be used as anything but a last resort. More often you can change how your state is structured/updated/read to avoid it.

**Comment by Fakhatova** [comment]
I am quite confused on what is outcome should look like on this one, but as I understand this issue is that number ''2''  is getting logged twice on UseEffect function. And this is caused because async/await is batching setState incorrectly. So I did mess around and figured that we are invoking setState within async/await function which is will fire setTimeout first and and it will move to que stack where it should wait 5 second before resolving, but meanwhile it will keep firing next line which is setState, so I did end up moving setState outside of async/await function, to let function resolve first before setting new state of array, and ready. let me know if I am completely understanding this wrong. And here is link to my solution. https://codesandbox.io/s/damp-hill-uhg6w?file=/src/App.tsx
